paddle/framework/ddim.cc

```diff
@@ -178,13 +179,56 @@ std::vector<int> vectorize(const DDim& ddim) {
   return result;
 }
 
+struct ProductVisitor : public boost::static_visitor<ssize_t> {
+  template <int D>
+  ssize_t operator()(const Dim<D>& dim) {
+    return product(dim);
+  }
+};
+
 ssize_t product(const DDim& ddim) {
-  ssize_t result = 1;
-  std::vector<int> v = vectorize(ddim);
-  for (auto i : v) {
-    result *= i;
+  ProductVisitor visitor;
+  return boost::apply_visitor(visitor, ddim);
```

Comment:
See https://github.com/Canpio/Paddle/wiki/Majel:-From-Allocation-to-DArray
In the last section, I show how `DArray` is transformed into `Array` in `boost::apply_visitor`. The principle is also suitable for `DDim` and `Dim`.